File: servers/python/routes/v1/comments.py
import logging
import sqlite3
from typing import Annotated

from anyio import Path
from fastapi import APIRouter, Depends, Response, status
from pydantic import BaseModel, field_validator

import constants
from routes import deps
from storage import Storage

logger = logging.getLogger("uvicorn.error")
router = APIRouter(prefix="/comments", dependencies=[Depends(deps.get_user_id)])


class UpdateCommentBody(BaseModel):
    text: str

    @field_validator("text")
    @classmethod
    def strip(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("INVALID")
        return v.strip()


@router.patch("/{comment_id}")
def update_comment(
    response: Response,
    comment_id: Annotated[int, Path()],
    body: UpdateCommentBody,
    db: sqlite3.Connection = Depends(deps.get_db),
    user_id=Depends(deps.get_user_id),
):
    try:
        db.execute("UPDATE comments SET text = ? WHERE id = ? AND user_id = ?", (body.text, comment_id, user_id))
        return {"success": True, "error": None}
    except Exception as e:
        logger.error(f"comments.update_comment: {e}")
        response.status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
        return {"success": False, "error": {"code": "INTERNAL_SERVER_ERROR"}}


@router.delete("/{comment_id}")
def delete_comment(
    response: Response,
    comment_id: Annotated[int, Path()],
    db: sqlite3.Connection = Depends(deps.get_db),
    storage: Storage = Depends(deps.get_storage),
    user_id=Depends(deps.get_user_id),
):
    try:
        attachments = db.execute("SELECT url FROM attachments WHERE comment_id = ?", (comment_id,)).fetchall()
        keys = [attachment["url"].removeprefix(constants.ASSETS_BASE_URL + "/") for attachment in attachments]

        db.execute("DELETE FROM comments WHERE id = ? AND user_id = ?", (comment_id, user_id))
        if len(keys) > 0:
            storage.delete(keys)

        return {"success": True, "error": None}
    except Exception as e:
        logger.error(f"comments.delete_comment: {e}")
        response.status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
        return {"success": False, "error": {"code": "INTERNAL_SERVER_ERROR"}}

File: servers/cloudflare-workers/src/handlers/v1/comments.ts
import { Hono } from 'hono';

import * as schemas from './schemas';

import * as validator from '../../validator';

const comments = new Hono<Env>();

comments.patch('/:id', validator.json(schemas.updateComment), async (c) => {
  const { text } = await c.req.valid('json');
  const stmt = c.env.DB.prepare('UPDATE comments SET text = ?3 WHERE id = ?1 AND user_id = ?2');
  await stmt.bind(c.req.param('id'), c.var.userId, text).run();
  return c.json({ success: true, error: null }, 200);
});

comments.delete('/:id', async (c) => {
  const commentId = c.req.param('id');

  const stmt = c.env.DB.prepare(`SELECT url FROM attachments WHERE id = ?`);
  const { results } = await stmt.bind(commentId).all<{ url: string }>();
  const keys = results.map((attachment) => attachment.url.replace(c.var.config.assetsBaseUrl + '/', ''));

  await Promise.all([
    c.env.DB.prepare('DELETE FROM comments WHERE id = ? AND user_id = ?').bind(commentId, c.var.userId).run(),
    c.env.Bucket.delete(keys),
  ]);

  return c.json({ success: true, error: null }, 200);
});

export default comments;
